File: app/src/main/java/com/example/messengerandroid/model/CharacterAdapter.java
package com.example.messengerandroid.model;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.messengerandroid.databinding.ItemCharacterBinding;

import java.util.List;

public class CharacterAdapter extends RecyclerView.Adapter<CharacterAdapter.CharacterViewHolder> {

    private final List<Character> characterList;

    public CharacterAdapter(List<Character> characters) {
        this.characterList = characters;
    }

    @NonNull
    @Override
    public CharacterViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ItemCharacterBinding binding = ItemCharacterBinding.inflate(
                LayoutInflater.from(parent.getContext()), parent, false
        );
        return new CharacterViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull CharacterViewHolder holder, int position) {
        Character character = characterList.get(position);
        holder.bind(character);
    }

    @Override
    public int getItemCount() {
        return characterList.size();
    }

    public static class CharacterViewHolder extends RecyclerView.ViewHolder {

        private final ItemCharacterBinding binding;

        public CharacterViewHolder(ItemCharacterBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        public void bind(Character character) {
            binding.tvName.setText(character.getName() == null || character.getName().isEmpty() ? "Unknown" : character.getName());
            binding.tvCulture.setText(character.getCulture() == null || character.getCulture().isEmpty() ? "Unknown" : character.getCulture());
            binding.tvBorn.setText(character.getBorn() == null || character.getBorn().isEmpty() ? "Unknown" : character.getBorn());
            binding.tvTitles.setText(character.getTitles().isEmpty() ? "Unknown" : String.join(", ", character.getTitles()));
            binding.tvAliases.setText(character.getAliases().isEmpty() ? "Unknown" : String.join(", ", character.getAliases()));
            binding.tvPlayedBy.setText(character.getPlayedBy().isEmpty() ? "Unknown" : String.join(", ", character.getPlayedBy()));
        }
    }
}

File: app/src/main/java/com/example/messengerandroid/model/CharacterAdapter.kt
package com.example.messengerandroid.model

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.example.messengerandroid.databinding.ItemCharacterBinding

class CharacterAdapter(
    private val characterList: List<Character>
) : RecyclerView.Adapter<CharacterAdapter.CharacterViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CharacterViewHolder {
        val binding = ItemCharacterBinding.inflate(
            LayoutInflater.from(parent.context), parent, false
        )
        return CharacterViewHolder(binding)
    }

    override fun onBindViewHolder(holder: CharacterViewHolder, position: Int) {
        val character = characterList[position]
        holder.bind(character)
    }

    override fun getItemCount(): Int = characterList.size

    class CharacterViewHolder(
        private val binding: ItemCharacterBinding
    ) : RecyclerView.ViewHolder(binding.root) {

        fun bind(character: Character) {
            binding.apply {
                tvName.text = character.name.takeIf { !it.isNullOrEmpty() } ?: "Unknown"
                tvCulture.text = character.culture.takeIf { !it.isNullOrEmpty() } ?: "Unknown"
                tvBorn.text = character.born.takeIf { !it.isNullOrEmpty() } ?: "Unknown"
                tvTitles.text = character.titles?.takeIf { it.isNotEmpty() }?.joinToString(", ") ?: "Unknown"
                tvAliases.text = character.aliases?.takeIf { it.isNotEmpty() }?.joinToString(", ") ?: "Unknown"
                tvPlayedBy.text = character.playedBy?.takeIf { it.isNotEmpty() }?.joinToString(", ") ?: "Unknown"
            }
        }
    }
}
